blog/2020/05/invalidations.md

```diff
@@ -0,0 +1,784 @@
+
+julia> mi.def
+(::Type{T})(x::Number) where T<:AbstractChar in Base at char.jl:48
+```
+
+`trees[end]` selected the last (most consequential) method and the invalidations it triggered; indexing this with `:backedges` selected the category (`:mt_backedges`, `:backedges`, or `:mt_cache`), and the integer index selected the particular entry from that category.
+This returns a pair `MethodInstance => InstanceTree`, where the latter is a type encoding the tree.
+We'll see how to work with `InstanceTree`s in a moment, for now we want to focus on the `mi` portion of that pair.
+
+You may find it surprising that this method signature is ambiguous with `(::Type{X})(x::Real) where X<:FixedPoint`: after all, an `AbstractChar` is quite different from a `FixedPoint` number.
+We can discover why with
+
+```julia-repl
+julia> tree.method.sig
+Tuple{Type{X},Real} where X<:FixedPoint
+
+julia> mi.specTypes
+Tuple{Type{T} where T<:AbstractChar,Int32}
+
+julia> typeintersect(tree.method.sig, mi.specTypes)
+Tuple{Type{Union{}},Int32}
+```
+
+These two signatures have non-empty intersection.
+The second parameter, `Int32`, makes sense as the intersection of `Int32` and `Real`.
+The first arises from
+
+```julia-repl
+julia> typeintersect(Type{<:FixedPoint}, Type{<:AbstractChar})
+Type{Union{}}
+```
+
+which shows that there is one Type, the "empty Type", that lies in their intersection.
+
+There are good reasons to believe that the right way to fix such methods is to exclude ambiguous pairs from invalidation--if it were to be called by the compiled code, it would trigger an error anyway.
+If this gets changed in Julia, then all the ones marked "ambiguous" should magically disappear.
+Consequently, we can turn our attention to other cases.
+
+Let's look at the next item up the list:
+
+```julia-repl
+julia> tree = trees[end-1]
+insert reduce_empty(::typeof(Base.mul_prod), ::Type{F}) where F<:FixedPoint in FixedPointNumbers at /home/tim/.julia/packages/FixedPointNumbers/w2pxG/src/FixedPointNumbers.jl:225 invalidated:
+   backedges: MethodInstance for reduce_empty(::Function, ::Type{T} where T) triggered MethodInstance for reduce_empty(::Base.BottomRF{typeof(max)}, ::Type{VersionNumber}) (136 children) more specific
+```
+
+`reduce_empty(::typeof(Base.mul_prod), ::Type{F}) where F<:FixedPoint` is strictly more specific than `reduce_empty(::Function, ::Type{T} where T)`.
+This might look like one of those "necessary" invalidations.
+However, even though it's marked "more specific," the new method `reduce_empty(::typeof(Base.add_sum), ::Type{F}) where F<:FixedPoint` can't be reached from a call `reduce_empty(::Base.BottomRF{typeof(max)}, ::Type{VersionNumber})`:
+
+```julia-repl
+julia> mi, node = tree[:backedges, 1]
+MethodInstance for reduce_empty(::Function, ::Type{T} where T) => MethodInstance for reduce_empty(::Base.BottomRF{typeof(max)}, ::Type{VersionNumber}) at depth 0 with 143 children
```

Comment:
I'm curious — why does this print this as a `MethodInstance for reduce_empty(::Base.BottomRF{typeof(max)}, ::Type{VersionNumber})` if Julia only specialized for `::Type{T} where T`?  I thought the `MethodInstance`s reflected the specializations.

Reply:
Interesting, I didn't know we sometimes specialized on `::Type{T}` without the explicit type-var.  That makes this make more sense to me now.